### PR TITLE
[WIP] New unified TextLayout

### PR DESCRIPTION
**Pull request by Gillibald**
- What does the pull request do?
It introduces an abstraction to the text layout and makes it possible to support the inlines and complex shaped text. 
- What is the current behavior?
Currently, there is no abstraction for the text layout and only some basic stuff is done.
- What is the updated/expected behavior with this PR?
The new text layout supports fallbacks for glyphs. That means if a glyph isn't present on the current typeface the layout tries to find one that has the glyph and uses it. That way you can mix typefaces in a text.
- How was the solution implemented (if it's not obvious)?
This is quite complex and I will add some comments to the code and try to explain how things work. In general, the layout splits a text into text lines and each line consists of text runs. Each run only contains glyphs that are present on one typeface and only has one specific set of font properties. 

The layout is able to perform line breaks if needed (TextWrapping) and also breaks line on `\r` etc. Text that is split by TextWrapping is only split at a whitespace. We can add some more advanced word breaking in the future.

This adds TextTrimming support for FormattedText

## Fixed issues
#1817
#1897
#1925
#2452

Initial work for: #1809

### CODE REVIEW

**Comment by Gillibald** [comment]
Still need to fix some issues with hit testing and add the right measurements for unicode characters.

**Comment by Gillibald** [comment]
I will copy over some bits of the inlines PR to work on the missing TextSpan properties. We should be able to merge that later.

**Comment by Gillibald** [comment]
```
    public class TextHitTestResult
    {
        /// <summary>
        /// Gets the first index within the hit region.
        /// </summary>
        public int TextPosition { get; set; }

        /// <summary>
        /// The number of text positions within the hit region. 
        /// </summary>
        /// <value>
        /// The number of text positions.
        /// </value>
        public int Length { get; set; }

        /// <summary>
        /// Gets the bounding box of the hit region.
        /// </summary>
        /// <value>
        /// The bounding box.
        /// </value>
        public Rect Bounds { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether the point is inside the bounds of the text.
        /// </summary>
        public bool IsInside { get; set; }

        /// <summary>
        /// Gets a value indicating whether the hit is on the trailing edge of the character.
        /// </summary>
        public bool IsTrailing { get; set; }
    }
```

**Comment by Gillibald** [comment]
DirectWrite
textlayout_d2d1

Skia
textlayout_skia

**Comment by Gillibald** [comment]
To test this on Linux you need to change your SkiaSharp.props
```
<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <PackageReference Include="SkiaSharp" Version="1.68.0" />
    <PackageReference Include="SkiaSharp.HarfBuzz" Version="1.68.0" />
    <PackageReference Condition="'$(IncludeLinuxSkia)' == 'true'" Include="SkiaSharp.NativeAssets.Linux" Version="1.68.0" />
    <PackageReference Condition="'$(IncludeLinuxSkia)' == 'true'" Include="HarfBuzzSharp.NativeAssets.Linux" Version="1.4.6.2" />
  </ItemGroup>
</Project>
```
And make these packages available:
`SkiaSharp.1.68.0.nupkg`
`SkiaSharp.NativeAssets.Linux.1.68.0.nupkg`
`SkiaSharp.HarfBuzz.1.68.0.nupkg`
`HarfBuzzSharp.1.4.6.2.nupkg`
`HarfBuzzSharp.NativeAssets.Linux.1.4.6.2.nupkg`

**Comment by Gillibald** [comment]
Looks like I have to wait for a bug fix in SkiaSharp until I can finish this.

https://github.com/mono/SkiaSharp/issues/715

File: src/Skia/Avalonia.Skia/Avalonia.Skia.csproj [hunk 1]
```diff
@@ -6,6 +6,9 @@
     <IncludeLinuxSkia>true</IncludeLinuxSkia>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Avalonia.Skia.csproj]: Should be removed

File: Avalonia.sln [hunk 2]
```diff
@@ -1663,7 +1663,6 @@ Global
 		{D49233F8-F29C-47DD-9975-C4C9E4502720}.AppStore|iPhoneSimulator.ActiveCfg = Debug|Any CPU
 		{D49233F8-F29C-47DD-9975-C4C9E4502720}.AppStore|iPhoneSimulator.Build.0 = Debug|Any CPU
 		{D49233F8-F29C-47DD-9975-C4C9E4502720}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{D49233F8-F29C-47DD-9975-C4C9E4502720}.Debug|Any CPU.Build.0 = Debug|Any CPU
```
Comment (Gillibald) [inline Avalonia.sln]: Revert

File: src/Skia/Avalonia.Skia/Text/SKTextLayout.cs [hunk 3]
```diff
@@ -0,0 +1,1508 @@
...
+        /// <summary>
+        /// Creates the text line metrics.
+        /// </summary>
+        /// <param name="textRuns">The text runs.</param>
+        /// <param name="length">Text length</param>
+        /// <returns></returns>
+        private static SKTextLineMetrics CreateTextLineMetrics(IEnumerable<SKTextRun> textRuns, out int length)
+        {
+            var width = 0.0f;
+
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: Remove new line

File: src/Skia/Avalonia.Skia/Text/SKTextLayout.cs [hunk 4]
```diff
@@ -0,0 +1,1508 @@
...
+                sizeY += textLine.LineMetrics.Size.Height;
+            }
+
+            Size = new Size(sizeX, sizeY);
+
+            return textLines;
+        }
+
+        /// <summary>
+        /// Creates text run with a specific text format.
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: Fix comment

File: src/Skia/Avalonia.Skia/Text/SKTextRun.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        /// </value>
+        public float Width { get; }
+
+        /// <summary>
+        /// Gets the drawing effect.
+        /// </summary>
+        /// <value>
+        /// The drawing effect.
+        /// </value>
+        public IBrush DrawingEffect { get; }
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Text/SKTextRun.cs]: Rename to ForegroundBrush

**Comment by Gillibald** [comment]
If someone wants to try this you need to build this first https://github.com/Gillibald/SkiaSharp/tree/feature/HarfBuzzAddUtfOverloads

**Comment by Gillibald** [comment]
[NugetDependencies.zip](https://github.com/AvaloniaUI/Avalonia/files/2750822/NugetDependencies.zip)

**Comment by Gillibald** [comment]
[NugetDependencies.zip](https://github.com/AvaloniaUI/Avalonia/files/2750830/NugetDependencies.zip)

File: src/Skia/Avalonia.Skia/Text/SKTextLayout.cs [hunk 6]
```diff
@@ -0,0 +1,1610 @@
...
+
+                if (glyphCount == 0)
+                {
+                    var codePoint = char.IsHighSurrogate(runText.Span[0])
+                                        ? char.ConvertToUtf32(runText.Span[0], runText.Span[1])
+                                        : runText.Span[0];
+
+                    typeface = SKFontManager.Default.MatchCharacter(codePoint);
+
+                    if (codePoint > sizeof(short))
```
Comment (kekekeks) [inline src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: You've probably meant short.MaxValue here

Comment (Gillibald) [reply src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: Thanks

**Comment by Gillibald** [comment]
I have refactored everything to use ReadOnlySpan<T> and only store a SKTextPointer struct in each SKTextRun to reference a section within the text. This is not polished and can probably be simplified. If you find something obvious let me know. I am open for performance improvements. In terms of memory consumption, this should be quite good already.

File: src/Skia/Avalonia.Skia/Text/SKTextLayout.cs [hunk 7]
```diff
@@ -0,0 +1,1592 @@
...
+
+                var result = shaper.Shape(buffer, _paint);
+
+                var glyphsIds = result.Codepoints.Select(x => (ushort)x).ToArray();
+
+                var points = result.Points;
+
+                var clusters = result.Clusters;
+
+                var glyphClusters = CreateGlyphClusters(text, textPointer, fontMetrics, glyphsIds, clusters, points);
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: Calculate the total width when glyph clusters are created

File: src/Skia/Avalonia.Skia/Text/SKTextLayout.cs [hunk 8]
```diff
@@ -0,0 +1,1592 @@
...
+
+                if (glyphCount == 0)
+                {
+                    var codePoint = char.IsHighSurrogate(runText[0])
+                                        ? char.ConvertToUtf32(runText[0], runText[1])
+                                        : runText[0];
+
+                    typeface = SKFontManager.Default.MatchCharacter(codePoint);
+
+                    if (codePoint > short.MaxValue)
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: This needs to compare against the ushort.MaxValue

File: src/Skia/Avalonia.Skia/Text/SKTextLayout.cs [hunk 9]
```diff
@@ -0,0 +1,1602 @@
...
+
+            using (var shaper = new SKShaper(textFormat.Typeface))
+            using (var buffer = new HarfBuzzSharp.Buffer())
+            {
+                buffer.AddUtf16(text, (uint)textPointer.StartingIndex, textPointer.Length);
+
+                buffer.GuessSegmentProperties();
+
+                var result = shaper.Shape(buffer, _paint);
+
```
Comment (Gillibald) [inline src/Skia/Avalonia.Skia/Text/SKTextLayout.cs]: Don't use the result and instead use the buffer directly to get all glyph advances etc

**Comment by Gillibald** [comment]
This is now depended on https://github.com/mono/SkiaSharp/pull/766

**Comment by Gillibald** [comment]
https://dev.azure.com/xamarin/public/_build/results?buildId=118&view=logs&jobId=3ce5e921-da7d-53ba-2ce2-68a0e19023ee

**Comment by Gillibald** [comment]
Optimization: Stop layout process if max width is reached and wrap = nowrap

**Comment by Gillibald** [comment]
In case somone wants to try this out I have uploaded current HarfBuzzSharp
[HarfBuzzSharp.zip](https://github.com/AvaloniaUI/Avalonia/files/3050779/HarfBuzzSharp.zip)

**Comment by grokys** [comment]
@Gillibald I've tried to review this PR but it's really hard to review because it introduces so many formatting changes unrelated to the functionality.

Would it be possible to undo the arbitrary changes so that it's possible to see what has changed functionally? I'd be happy to submit a PR to this PR doing so if you're busy.

**Comment by Gillibald** [comment]
Would help me a lot if you could do that for me. Sry for all the unrelated changes. The TextBox.cs is a mess. I should really disable all auto formatting.

TextBox itself has changed a lot. Sadly I didn't know how to write unit tests for it. A lot of the functionality is executed by events. 

Is the a tool I can work with to revert these formatting changes?

I know this will be hard to review just let me know if something isn't clear to you.

**Comment by Gillibald** [comment]
Text hit testing now supports clusters of glyphs. That way selection and navigation works properly for sequences of codepoints that are treated as one unit. Emojis for example are represented as a surrogate pair and should be removed or navigated to as a pair. More complex sequences like ligatures are also handled properly.

**Comment by Gillibald** [comment]
I have removed FontSize from Typeface to support the reuse of it at different font sizes.

We can fix all the formatting, naming etc in a different PR. It just felt right to fix it when I touched parts of the code base but makes it hard to review. My bad.

**Comment by grokys** [comment]
> The TextBox.cs is a mess. Sadly I didn't know how to write unit tests for it. A lot of the functionality is executed by events.

Yeah, I agree - it really needs an overhaul and proper unit tests...

> Is there a tool I can work with to revert these formatting changes?

Unfortunately I don't know of any. I'll do it manually.

> We can fix all the formatting, naming etc in a different PR. It just felt right to fix it when I touched parts of the code base but makes it hard to review. My bad.

Yeah, and to an extent it's ok to do that with smaller PRs. However for bigger PRs like this, it's best to avoid formatting changes/refactors if possible. 

Anyway, not a problem - I'll submit a PR which targets this PR to remove the formatting changes -- it'll be useful as a way of reviewing the PR even.

**Comment by Gillibald** [comment]
To tests this artifacts can be downloaded here: https://dev.azure.com/xamarin/public/_build/results?buildId=422

**Comment by Gillibald** [comment]
I could revert most of the formatting changes. Should be easier to review now.

**Comment by mat1jaczyyy** [comment]
Any plans on merging this soon? I would love to utilize the TextTrimming property this introduces.

**Comment by Gillibald** [comment]
This is almost ready but still needs some work. I hope I can find some time this week to fix glyph fallbacks for missing fonts.

File: src/Avalonia.Visuals/Media/GlyphRun.cs [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                }
+            }
+
+            return new Rect(0, 0, width, height);
+        }
+
+        private void Set<T>(ref T field, T value)
+        {
+            field = value;
+            _glyphRunImpl?.Dispose();
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/GlyphRun.cs]: This is wrong. GlyphRun needs to be immutable once IGlyphRunImpl is created.

File: src/Avalonia.Visuals/Media/Text/FontMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+namespace Avalonia.Media.Text
+{
+    public readonly struct FontMetrics
+    {
+        private FontMetrics(double ascent, double descent, double lineGap, double underlineThickness,
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/Text/FontMetrics.cs]: Change this to except a GlyphTypeface and a font size.

File: src/Avalonia.Visuals/Media/Text/GlyphCluster.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Avalonia.Media.Text
+{
+    public class GlyphCluster
+    {
+        public GlyphCluster(int textPosition, int length, Rect bounds)
+        {
+            TextPosition = textPosition;
+            Length = length;
+            Bounds = bounds;
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/Text/GlyphCluster.cs]: Bounds - > Advance

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 13]
```diff
@@ -219,12 +223,10 @@ internal void CaretIndexChanged(int caretIndex)
                     _caretBlink = true;
                     _caretTimer.Stop();
                     _caretTimer.Start();
-                    InvalidateVisual();
```
Comment (Gillibald) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs]: Revert this

**Comment by kronic** [comment]
@Gillibald Any news?

**Comment by Gillibald** [comment]
Currently I am working on a minor TextBox rework to support text editing in advanced scenarios. Static text works just fine.

**Comment by aguahombre** [comment]
Any idea when this PR will get merges back into the master?
I would like to port some WPF code that uses Inlines for hyperlinks within a TextBlock
